### PR TITLE
Support uploading files with Size filtering and AQL avoiding

### PR DESCRIPTION
**Pull request by sverdlov93**
- [ ] All [tests](https://github.com/jfrog/jfrog-client-go#tests) passed. If this feature is not already covered by the tests, I added new tests.
- [ ] All [static analysis checks](https://github.com/jfrog/jfrog-client-go/actions/workflows/analysis.yml) passed.
- [ ] This pull request is on the dev branch.
- [ ] I used gofmt for formatting the code before submitting the pull request.
-----

- Added Sha256 and Size fields to DownloadParams to avoid fetching them with AQL
- Following this change added support for checksum validation for sha256 and united the checksum validation code.
- Added SizeThreshold field to support filtering files by size.
- Improved the ListFiles func, before it was first collecting all files and then running on them again to filter. Now it's filtering the files while collecting them.
- Also added support for filtering by size threshold.

### CODE REVIEW

File: utils/io/fileutils/files.go [hunk 1]
```diff
@@ -164,8 +164,36 @@ func ListFilesRecursiveWalkIntoDirSymlink(path string, walkIntoDirSymlink bool)
...
+}
+
+// Return the recursive list of files and directories in the specified path
+func listFilesRecursiveWithFilterFunc(path string, walkIntoDirSymlink bool, filterFunc func(filePath string) (bool, error)) (fileList []string, err error) {
+	fileList = []string{}
+	err = gofrog.Walk(path, func(path string, f os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		satisfy, err := filterFunc(path)
```
Comment (yahavi) [inline utils/io/fileutils/files.go:182]:
```suggestion
		include, err := filterFunc(path)
```

File: utils/io/fileutils/files.go [hunk 2]
```diff
@@ -164,8 +164,36 @@ func ListFilesRecursiveWalkIntoDirSymlink(path string, walkIntoDirSymlink bool)
...
+	err = errorutils.CheckError(err)
+	return
+}
+
 // Return all files in the specified path who satisfy the filter func. Not recursive.
-func ListFilesByFilterFunc(path string, filterFunc func(filePath string) (bool, error)) ([]string, error) {
+func listFilesNonRecursiveWithFilterFunc(path string, includeDirs bool, filterFunc func(filePath string) (bool, error)) ([]string, error) {
 	sep := GetFileSeparator()
 	if !strings.HasSuffix(path, sep) {
 		path += sep
```
Comment (yahavi) [inline utils/io/fileutils/files.go:199]: Ignored error in line 202

Comment (yahavi) [reply utils/io/fileutils/files.go:199]: The path in line 206 is concatenated with + and not with filepath.Join

File: utils/io/fileutils/files.go [hunk 3]
```diff
@@ -164,8 +164,36 @@ func ListFilesRecursiveWalkIntoDirSymlink(path string, walkIntoDirSymlink bool)
...
+		}
+		return nil
+	}, walkIntoDirSymlink)
+	err = errorutils.CheckError(err)
+	return
+}
+
 // Return all files in the specified path who satisfy the filter func. Not recursive.
-func ListFilesByFilterFunc(path string, filterFunc func(filePath string) (bool, error)) ([]string, error) {
+func listFilesNonRecursiveWithFilterFunc(path string, includeDirs bool, filterFunc func(filePath string) (bool, error)) ([]string, error) {
```
Comment (yahavi) [inline utils/io/fileutils/files.go:196]: Can we unite listFilesRecursiveWithFilterFunc and listFilesNonRecursiveWithFilterFunc to use only gofrog.Walk?

File: artifactory/services/download.go [hunk 4]
```diff
@@ -197,9 +204,50 @@ func (ds *DownloadService) prepareTasks(producer parallel.Runner, expectedChan c
...
+		Size:   downloadParams.Size,
+		Sha256: downloadParams.Sha256,
+	}
+	writer.Write(*resultItem)
+	return content.NewContentReader(writer.GetFilePath(), writer.GetArrayKey()), nil
+}
+
+func breakFileDownloadPathToParts(downloadPath string) (repo, path, name string, err error) {
+	if utils.IsWildcardPattern(downloadPath) {
+		return "", "", "", errors.New("downloading without AQL is not supported for the provided wildcard pattern: " + downloadPath)
```
Comment (yahavi) [inline artifactory/services/download.go:238]:
```suggestion
		return "", "", "", errorutils.CheckErrorf("downloading without AQL is not supported for the provided wildcard pattern: " + downloadPath)
```

File: artifactory/services/fspatterns/utils.go [hunk 5]
```diff
@@ -53,23 +49,33 @@ func PrepareExcludePathPattern(exclusions []string, patternType utils.PatternTyp
...
-		if files[i] == "." {
-			continue
+// Returns a function that filters files according to the provided parameters
+func filterFilesFunc(rootPath string, excludePathPattern string, sizeThreshold *SizeThreshold) func(filePath string) (included bool, err error) {
+	return func(path string) (included bool, err error) {
+		if path == "." {
+			return false, nil
 		}
-		excludedPath, err = isPathExcluded(strings.TrimPrefix(files[i], rootPath), excludePathPattern)
+		excludedPath, err := isPathExcluded(strings.TrimPrefix(path, rootPath), excludePathPattern)
```
Comment (yahavi) [inline artifactory/services/fspatterns/utils.go:58]: I think that isPathExcluded does not do checkerror. Let's add it here or there.

File: artifactory/services/fspatterns/utils.go [hunk 6]
```diff
@@ -53,23 +49,33 @@ func PrepareExcludePathPattern(exclusions []string, patternType utils.PatternTyp
...
-			log.Debug(fmt.Sprintf("The path '%s' is excluded", files[i]))
+		if excludedPath {
+			log.Debug(fmt.Sprintf("The path '%s' is excluded", path))
+			return false, nil
+		}
+
+		if sizeThreshold != nil {
+			fileInfo, err := os.Stat(path)
+			if err != nil {
+				return false, err
```
Comment (yahavi) [inline artifactory/services/fspatterns/utils.go:70]:
```suggestion
				return false, errorutils.CheckError(err)
```

File: artifactory/services/download.go [hunk 7]
```diff
@@ -164,17 +164,24 @@ func (ds *DownloadService) prepareTasks(producer parallel.Runner, expectedChan c
...
-			// Search items.
-			switch downloadParams.GetSpecType() {
-			case utils.WILDCARD:
-				reader, err = ds.collectFilesUsingWildcardPattern(downloadParams)
-			case utils.BUILD:
-				reader, err = utils.SearchBySpecWithBuild(downloadParams.GetFile(), ds)
-			case utils.AQL:
-				reader, err = utils.SearchBySpecWithAql(downloadParams.GetFile(), ds, utils.SYMLINK)
+
+			if downloadParams.Sha256 != "" {
```
Comment (yahavi) [inline artifactory/services/download.go:172]: Let's add this new feature to the README

File: artifactory/services/upload.go [hunk 8]
```diff
@@ -718,6 +719,7 @@ type UploadParams struct {
 	Archive              string
 	// When using the 'archive' option for upload, we can control the target path inside the uploaded archive using placeholders. This operation determines the TargetPathInArchive value.
 	TargetPathInArchive string
+	SizeLimit           *fspatterns.SizeThreshold
```
Comment (yahavi) [inline artifactory/services/upload.go:722]: Let's add this new feature to the README

**Review by yahavi — CHANGES_REQUESTED** [review]

File: utils/io/fileutils/files.go (hunk 3 above)
Comment (sverdlov93) [reply utils/io/fileutils/files.go:196]: @yahavi How can we use gofrog.Walk not recursive?

File: artifactory/services/fspatterns/sizethreshold.go [hunk 9]
```diff
@@ -0,0 +1,27 @@
+package fspatterns
```
[the remark this answers is not in the document]
Comment (sverdlov93) [reply artifactory/services/fspatterns/sizethreshold.go:1]: @yahavi cant see the benefit here?

[the remark this answers is not in the document]
Comment (yahavi) [reply artifactory/services/fspatterns/sizethreshold.go:1]: Replacing this code with inheritance can provide a few benefits:

1. Simplification and Clarity: By using inheritance, each specific condition can be encapsulated in its own class, making the code easier to understand and maintain.
2. Eliminates Switch Statements: The switch statement in IsSizeWithinThreshold is replaced by polymorphism, allowing each condition to handle the size check in its own way.
3. Extensibility: Adding new threshold conditions can be done by simply adding new subclasses, without modifying existing code. This follows the Open/Closed Principle, which states that classes should be open for extension but closed for modification.

**Review by yahavi — CHANGES_REQUESTED** [review]

File: utils/io/fileutils/files.go (hunk 3 above)
Comment (yahavi) [reply utils/io/fileutils/files.go:196]: I'm not sure. Something like this?

```go
	err = gofrog.Walk(rootPath, func(path string, f os.FileInfo, err error) error {
		if err != nil {
			return err
		}
                // ...
		if nonRecursive && rootPath != path {
			return gofrog.ErrSkipDir
		}
		return nil
	}, walkIntoDirSymlink)
```

File: artifactory/services/fspatterns/sizethreshold.go [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+
+const (
+	// GreaterThan is greater & equal
+	GreaterEqualThan ThresholdCondition = iota
+	// LessThan is only less
+	LessThan
+)
+
+type SizeThreshold struct {
+	Size      int64
```
Comment (yahavi) [inline artifactory/services/fspatterns/sizethreshold.go:14]:
```suggestion
	SizeInBytes      int64
```

File: utils/io/fileutils/files.go (hunk 3 above)
Comment (sverdlov93) [reply utils/io/fileutils/files.go]: Gaon @yahavi

File: README.md [hunk 11]
```diff
@@ -482,7 +485,9 @@ params.Retries = 5
 params.SplitCount = 2
 // MinSplitSize default value: 5120
 params.MinSplitSize = 7168
-
+// Optional fields to avoid AQL request
+Sha256 = "aaaaabbbbbcccccc"
```
Comment (yahavi) [inline README.md:489]: Let's use a real sha256 to make no room for mistakes
```suggestion
Sha256 = "5feceb66ffc86f38d952786c6d696c79c2dbc239dd4e91b46729d73a27fb57e9"
```

File: utils/io/fileutils/files.go (hunk 3 above)
Comment (sverdlov93) [reply utils/io/fileutils/files.go]: Cancel the Gaon, it was wrong 😞 
the files on the top level are not ==rootPath
they are rootPath/file.
@yahavi

**Pull request merged**